File: DSA/Java/src/dsa/lib/math/ds/Triangle.java
package dsa.lib.math.ds;


import java.util.List;

public class Triangle {
    // LeetCode120
    //DFS
    public int minimumTotal(List<List<Integer>> triangle) {
        return dfs(triangle, 0, 0);
    }

    int dfs(List<List<Integer>> triangle, int i, int j) {
        if (i >= triangle.size()) return 0;

        int downLeft = dfs(triangle, i + 1, j);
        int downRight = dfs(triangle, i + 1, j + 1);

        return triangle.get(i).get(j) + Math.min(downLeft, downRight);
    }

    //DFS with memorization
    public int minimumTotal1(List<List<Integer>> triangle) {
        final int m = triangle.size();
        Integer[][] memo = new Integer[m][m];
        return dfs1(triangle, 0, 0, memo);
    }

    int dfs1(List<List<Integer>> triangle, int i, int j, Integer[][] memo) {
        if (i >= triangle.size()) return 0;
        if (memo[i][j] != null) return memo[i][j];

        int downLeft = dfs1(triangle, i + 1, j, memo);
        int downRight = dfs1(triangle, i + 1, j + 1, memo);
        memo[i][j] = triangle.get(i).get(j) + Math.min(downLeft, downRight);

        return memo[i][j];
    }

    //    Dynamic Programming
    public int minimumTotal2(List<List<Integer>> triangle) {
        // dp initialize
        int m = triangle.size();
        int[][] dp = new int[m][m];
        copy(triangle.get(m - 1), 0, dp[m - 1], 0, triangle.get(m - 1).size());

        // dp formula
        for (int i = m - 2; i >= 0; i--) {
            int n = triangle.get(i).size();
            for (int j = 0; j < n; j++) {
                dp[i][j] = triangle.get(i).get(j) + Math.min(dp[i + 1][j], dp[i + 1][j + 1]);
            }
        }

        // dp result
        return dp[0][0];
    }

    //    dp with Space Optimized
    public int minimumTotal3(List<List<Integer>> triangle) {
        // dp initialize
        int m = triangle.size();
        int[] dp = new int[m];
        copy(triangle.get(m - 1), 0, dp, 0, triangle.get(m - 1).size());

        // dp formula
        for (int i = m - 2; i >= 0; i--) {
            int n = triangle.get(i).size();
            for (int j = 0; j < n; j++) {
                dp[j] = triangle.get(i).get(j) + Math.min(dp[j], dp[j + 1]);
            }
        }

        // dp result
        return dp[0];
    }

    public static void copy(List<Integer> src, int srcPos, int[] des, int desPos, int n) {
        for (int i = 0; i < n; ++i) des[desPos + i] = src.get(srcPos + i);
    }

}

File: DSA/Go/dsa/lib/math/ds/triangle.go
package ds

import (
	dsa "dsa/lib/utils"
	"math"
)

// LeetCode120
// dfs. Redundant computation
func minimumTotal(triangle [][]int) int {
	return dfs(triangle, 0, 0)
}

func dfs(triangle [][]int, i int, j int) int {
	if i >= len(triangle) {
		return 0
	}

	downLeft := dfs(triangle, i+1, j)
	downRight := dfs(triangle, i+1, j+1)
	return triangle[i][j] + dsa.Min(downLeft, downRight)
}

// dfs with memorization
func minimumTotal1(triangle [][]int) int {
	m := len(triangle)
	memo := dsa.CreateMatrix[int](m, m)
	dsa.FillMatrix(memo, math.MinInt)

	return dfs1(triangle, 0, 0, memo)
}

func dfs1(triangle [][]int, i int, j int, memo [][]int) int {
	if i >= len(triangle) {
		return 0
	}

	if memo[i][j] != math.MinInt {
		return memo[i][j]
	}

	downLeft := dfs(triangle, i+1, j)
	downRight := dfs(triangle, i+1, j+1)
	memo[i][j] = triangle[i][j] + dsa.Min(downLeft, downRight)

	return memo[i][j]
}

// Dynamic Programming. Bottom up
func minimumTotal2(triangle [][]int) int {
	m := len(triangle)
	dp := dsa.CreateMatrix[int](m, m)

	// dp initialize
	copy(dp[m-1], triangle[m-1])

	// dp formula
	for i := m - 2; i >= 0; i-- {
		n := len(triangle[i])
		for j := 0; j < n; j++ {
			dp[i][j] = triangle[i][j] + dsa.Min(dp[i+1][j], dp[i+1][j+1])
		}
	}

	// result
	return dp[0][0]
}

// Dynamic Programming. Bottom up. Space Optimized
func minimumTotal3(triangle [][]int) int {
	m := len(triangle)
	dp := make([]int, m)

	// dp initialize
	copy(dp, triangle[m-1])

	// dp formula
	for i := m - 2; i >= 0; i-- {
		n := len(triangle[i])
		for j := 0; j < n; j++ {
			dp[j] = triangle[i][j] + dsa.Min(dp[j], dp[j+1])
		}
	}

	// result
	return dp[0]
}
